csrc/id_model/id_graph.cpp

```diff
@@ -0,0 +1,555 @@
+    return;
+  }
+
+  // Expr inputs are mapped. If propagate_exprs_ is true, map the
+  // exprs and outputs
+  if (propagate_through_exprs_) {
+    mapExprs(expr0, expr1);
+    mapThroughExpr(expr0, expr1, forward);
+  } else if (
+      inputGroups(toGroup(expr0)) == inputGroups(toGroup(expr1)) &&
+      outputGroups(toGroup(expr0)) == outputGroups(toGroup(expr1))) {
+    mapExprs(expr0, expr1);
+  }
+}
+
+void IdGraph::mapExprs(Expr* expr0, Expr* expr1) {
+  if (expr0 == expr1) {
+    return;
+  }
+
+  if (disjointExprSets().strictAreMapped(expr0, expr1)) {
+    return;
+  }
+
+  const ExprGroup& expr0_orig_group = toGroup(expr0);
+  const ExprGroup& expr1_orig_group = toGroup(expr1);
+
+  disjointExprSets().mapEntries(expr0, expr1);
+
+  const ExprGroup& expr_new_group = toGroup(expr0);
+
+  // Update unique uses of producers
+  IdGroups producers;
+  for (auto expr : std::vector<Expr*>{expr0, expr1}) {
+    for (auto input_id : ir_utils::filterByType<IterDomain>(expr->inputs())) {
+      producers.pushBack(toGroup(input_id));
+    }
+  }
+
+  for (const IdGroup& producer_group : producers) {
+    unique_uses_.at(producer_group).erase(expr0_orig_group);
+    unique_uses_.at(producer_group).erase(expr1_orig_group);
+    unique_uses_.at(producer_group).pushBack(expr_new_group);
+  }
+
+  // Update unique definitinos of consumers
+  IdGroups consumers;
+  for (auto expr : std::vector<Expr*>{expr0, expr1}) {
+    for (auto output_id : ir_utils::filterByType<IterDomain>(expr->outputs())) {
+      consumers.pushBack(toGroup(output_id));
+    }
+  }
+
+  for (const IdGroup& consumer_group : consumers) {
+    unique_definitions_.at(consumer_group).erase(expr0_orig_group);
+    unique_definitions_.at(consumer_group).erase(expr1_orig_group);
+    unique_definitions_.at(consumer_group).pushBack(expr_new_group);
+  }
+}
+
+bool IdGraph::mapThroughExpr(Expr* first, Expr* second, bool forward) {
+  if (first == nullptr || second == nullptr) {
+    return false;
+  }
+
+  if (!exprsMap(first, second, forward)) {
+    return false;
+  }
+
+  NVF_ERROR(
+      propagate_through_exprs_,
+      "Asked to propagate expression mappings on a graph that has propagate_exprs_ disabled.");
+
+  auto first_ids = ir_utils::filterByType<IterDomain>(
+                       forward ? first->outputs() : first->inputs())
+                       .vector();
+  auto second_ids = ir_utils::filterByType<IterDomain>(
+                        forward ? second->outputs() : second->inputs())
+                        .vector();
+  NVF_ERROR(
+      first_ids.size() == second_ids.size(),
+      "This should be unreachable, if transformation expressions match, their number of inputs and outputs should as well.\n However found:\n",
+      first->toString(),
+      "\nand\n",
+      second->toString());
+  for (auto out_i : c10::irange(first_ids.size())) {
+    mapIds(first_ids[out_i], second_ids[out_i]);
+  }
+
+  return true;
+}
+
+void IdGraph::mapThroughLoopSwizzles() {
```

Comment:
Just a heads up: We are not using loop swizzle any more in our matmul scheduler. I don't have an idea on whether loop swizzle will be useful in the future, or it will just be abandoned. So I don't recommend working hard to make loop swizzle working in id graph.